addons/project/models/project_task.py

```diff
@@ -722,41 +722,74 @@ def copy_data(self, default=None):
-            else:
-                vals['child_ids'] = [child.copy({'name': child.name}).id for child in task.child_ids]
+
+            vals['stage_id'] = task.stage_id.id
+            vals['name'] = task.name if self.env.context.get('copy_project') else _("%s (copy)", task.name)
             if task.recurrence_id:
                 vals['recurrence_id'] = task.recurrence_id.copy().id
+            if task.allow_milestones:
+                vals['milestone_id'] = milestone_mapping.get(vals['milestone_id'], vals['milestone_id'])
+            if task.child_ids:
+                default = {
+                    'depend_on_ids': False,
+                    'dependent_ids': False,
+                    'child_ids': False,
+                    'parent_id': False,
+                }
+                vals['child_ids'] = [Command.create(child_id.copy_data(default)[0]) for child_id in task.child_ids]
         return vals_list
 
+    def create_task_mapping(self, copied_tasks):
+        """
+        Thanks to the way create and command.create is handled, when a task with 2 children is copied, we have the guarantee that the children of the
+        copied task will have the same index in the child_ids recordset. We can use this behavior to create a mapping containing all the original tasks and their copy.
+        :return:
+            task_mapping: a dict containing the mapping of the original task ids and their copied task (k: original_task.id, v: new_task)
+            task_dependencies: a dict containing the ids of the dependencies of the original task when they have one.
+            (k: original_task_id, v: [original_task.depend_on_ids.ids, original_task.dependent_ids.ids]
+        """
+        task_mapping, task_dependencies = {}, {}
+        for original_task, copied_task in zip(self, copied_tasks):
+            task_mapping[original_task.id] = copied_task
+            if original_task.allow_task_dependencies and (original_task.depend_on_ids or original_task.dependent_ids):
+                task_dependencies[original_task.id] = [original_task.depend_on_ids.ids, original_task.dependent_ids.ids]
+            if original_task.child_ids:
+                # If the task has children, we have to call the method create_task_mapping to get their ids and dependencies mapping too.
+                children_mapping, children_dependencies = original_task.child_ids.create_task_mapping(copied_task.child_ids)
+                task_mapping.update(children_mapping)
+                task_dependencies.update(children_dependencies)
+        return task_mapping, task_dependencies
+
     def copy(self, default=None):
-        if 'task_mapping' not in self.env.context:
-            self = self.with_context(task_mapping={})
+        default = default or {}
+        default.update({
+            'depend_on_ids': False,
+            'dependent_ids': False,
+            'child_ids': False,
```

Comment:
I did not catch why we need to set those fields to false. 🤔

Reply:
Its a performance choice. If we let the depend fields be copied with their original value, its also going to trigger a write on the original tasks to add new dependencies. Since we have to later check the dependencies of the newly created task to set the correct ids, we will re-trigger a write on those original tasks to remove the needlessly added dependency. That's why the master copy() has those strange new_task.link() and old_task.unlink() on dependency fields.

e.a. 
task A, with depend_on_ids = task B.
[task A, task B].copy()
if we dont remove the depend_on_ids :
task A(copy), with depend_on_ids = task B
task B(copy), with dependent_ids = task A
=> this trigger a write on the original task A & B.
But since those tasks have been copied, we want to set the copied version in the dependencies of task A(copy) and task B (copy), not the original.
Hence, we have to re-trigger a write on the original task to remove the added dependencies.

Unlike the child_ids, we have no hierarchy in the dependencies, hence, we are force to handle this use case after the batch_create, and i did not find a better way than to not write any useless thing in the create to handle it after.

About the child_ids, i think it was a leftover of the draft of the algorithm. I'll try removing it and see if tests are failing

Reply:
edit : after some testing, it looks like the child_ids was indeed unneeded. I removed it, and the runbot is green.